saleor/warehouse/management.py

```diff
@@ -499,3 +504,89 @@ def _create_preorder_allocation(
         ),
         None,
     )
+
+
+@traced_atomic_transaction()
+def deactivate_preorder_for_variant(product_variant: ProductVariant):
+    """Complete preorder for product variant.
+
+    All preorder settings should be cleared and all preorder allocations
+    should be replaced by regular allocations.
+    """
+    if not product_variant.is_preorder:
+        return
+    channel_listings = ProductVariantChannelListing.objects.filter(
+        variant_id=product_variant.pk
+    )
+    channel_listings_pk = (channel_listing.id for channel_listing in channel_listings)
+    preorder_allocations = PreorderAllocation.objects.filter(
+        product_variant_channel_listing_id__in=channel_listings_pk
+    ).select_related("order_line", "order_line__order")
+
+    allocations_to_create = []
+    stocks_to_create = []
+    for preorder_allocation in preorder_allocations:
+        stock = _get_stock_for_preorder_allocation(preorder_allocation, product_variant)
+        if stock._state.adding:
+            stocks_to_create.append(stock)
+        allocations_to_create.append(
+            Allocation(
+                order_line=preorder_allocation.order_line,
+                stock=stock,
+                quantity_allocated=preorder_allocation.quantity,
+            )
+        )
+
+    if stocks_to_create:
+        Stock.objects.bulk_create(stocks_to_create)
+
+    if allocations_to_create:
+        Allocation.objects.bulk_create(allocations_to_create)
+
+    if preorder_allocations:
+        preorder_allocations.delete()
+
+    product_variant.preorder_global_threshold = None
+    product_variant.preorder_end_date = None
+    product_variant.is_preorder = False
+    product_variant.save(
+        update_fields=["preorder_global_threshold", "preorder_end_date", "is_preorder"]
+    )
+
+    for channel_listing in channel_listings:
+        channel_listing.preorder_quantity_threshold = None
+    ProductVariantChannelListing.objects.bulk_update(
+        channel_listings, ["preorder_quantity_threshold"]
+    )
+
+
+def _get_stock_for_preorder_allocation(
+    preorder_allocation: PreorderAllocation, product_variant: ProductVariant
+) -> Stock:
+    """Return stock where preordered variant should be allocated."""
+    order = preorder_allocation.order_line.order
+    shipping_method_id = order.shipping_method_id
+    if shipping_method_id is not None:
+        warehouse = Warehouse.objects.filter(
+            shipping_zones__id=order.shipping_method.shipping_zone_id  # type: ignore
+        ).first()
+    else:
+        from ..order.utils import get_order_country
+
+        country = get_order_country(order)
+        warehouse = Warehouse.objects.filter(
+            shipping_zones__countries__contains=country
+        ).first()
+
+    if not warehouse:
+        raise PreorderAllocationError(preorder_allocation.order_line)
```

Comment:
It could be useful to include a message explaining why the error happened, that would be later returned in the mutation response. It's easier to debug than just a generic message that we have now. If we had more `PreorderAllocationError` I would consider also adding a dedicated error code, but for now, just a message would do.

Reply:
This `PreorderAllocationError` was created similar way `InsufficientStock` works - it takes parameter and the proper message is prepared in constructor. However, you're right it's not exposed in API and it's bad - so I've changed it and the message with data from this line will be returned in API.